src/transforms/sort.js

```diff
@@ -0,0 +1,189 @@
+        description: [
+            'Sets the sort transform order.'
+        ].join(' ')
+    }
+};
+
+exports.supplyDefaults = function(transformIn) {
+    var transformOut = {};
+
+    function coerce(attr, dflt) {
+        return Lib.coerce(transformIn, transformOut, exports.attributes, attr, dflt);
+    }
+
+    var enabled = coerce('enabled');
+
+    if(enabled) {
+        coerce('target');
+        coerce('order');
+    }
+
+    return transformOut;
+};
+
+exports.calcTransform = function(gd, trace, opts) {
+    if(!opts.enabled) return;
+
+    var target = opts.target;
+    var targetArray = getTargetArray(trace, target);
+    var len = targetArray.length;
+
+    if(!len) return;
+
+    var arrayAttrs = PlotSchema.findArrayAttributes(trace);
+    var d2c = getDataToCoordFunc(gd, trace, target, targetArray);
+    var indices = getIndices(opts, targetArray, d2c);
+
+    for(var i = 0; i < arrayAttrs.length; i++) {
+        var np = Lib.nestedProperty(trace, arrayAttrs[i]);
+        var arrayCopy = np.get().slice();
+        var arrayNew = new Array(len);
+
+        for(var j = 0; j < len; j++) {
+            arrayNew[j] = arrayCopy[indices[j]];
+        }
+
+        np.set(arrayNew);
+    }
+};
+
+// TODO reuse for filter.js
+function getTargetArray(trace, target) {
+    if(typeof target === 'string' && target) {
+        var array = Lib.nestedProperty(trace, target).get();
+
+        return Array.isArray(array) ? array : [];
+    }
+    else if(Array.isArray(target)) return target.slice();
+
+    return false;
+}
+
+// TODO reuse for filter.js
+function getDataToCoordFunc(gd, trace, target, targetArray) {
+    var ax;
+
+    // If target points to an axis, use the type we already have for that
+    // axis to find the data type. Otherwise use the values to autotype.
+    if(target === 'x' || target === 'y' || target === 'z') {
+        ax = axisIds.getFromTrace(gd, trace, target);
+    }
+    // In the case of an array target, make a mock data array
+    // and call supplyDefaults to the data type and
+    // setup the data-to-calc method.
+    else if(Array.isArray(target)) {
+        ax = {
+            type: autoType(targetArray),
+            // TODO does this still work with the new hash object
+            _categories: []
+        };
+        setConvert(ax);
+
+        // build up ax._categories (usually done during ax.makeCalcdata()
+        if(ax.type === 'category') {
+            for(var i = 0; i < targetArray.length; i++) {
+                ax.d2c(targetArray[i]);
+            }
+        }
+    }
+
+    // if 'target' has corresponding axis
+    // -> use setConvert method
+    if(ax) return ax.d2c;
+
+    // special case for 'ids'
+    // -> cast to String
+    if(target === 'ids') return function(v) { return String(v); };
+
+    // otherwise (e.g. numeric-array of 'marker.color' or 'marker.size')
+    // -> cast to Number
+    return function(v) { return +v; };
+}
+
+function getIndices(opts, targetArray, d2c) {
```

Comment:
The function `d2c` seems to always be `getDataToCoordFunc` which looks a bit expensive. Since it's a lot of calls to this function - 2 calls on each comparison - there's probably benefit to first map (e.g. in a loop) to get the `d2c` values for all items, and then sort based on that. One way is sorting value/index pairs as above, or more simply, just the indices in a numeric array. Another is to add the `d2c` value as a property to the objects to be sorted, and then use `return a.d2cValue - b.d2cValue || a.originalIndex - b.originalIndex` similar to the above link in which case the objects themselves are easily sorted (using the accessor is very fast).